### ISSUE TITLE
segmentation fault modifying array ref during push

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9776]
Migrated from [rt.perl.org#66930](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=66930) (status was 'open')

Searchable as RT66930$

**Comment by p5pRT** [comment]
### From dan@gina.alaska.edu

#### Created by dan@gina.alaska.edu



The following program causes a segmentation fault&#8203;:

\#\!/usr/bin/perl 

use strict;
use warnings;

my $x;
\#my $x = \[\];  \# also crashes

push @&#8203;$x\, abc\(\);

sub abc \{
&nbsp;    $x = 1;
&nbsp;    \#$x = \[\];   \# also crashes
&nbsp;    return 2;
&nbsp;    \#return $x = 1;   \# also crashes
&nbsp;    \# absense of return statement prevents crash
\}

\#\#\#\#\#

$ gdb \-\-args perl crash\.pl 
Program received signal SIGSEGV\, Segmentation fault\.
0x00007ffff7d091a8 in Perl\_av\_store \(my\_perl=0x602010\, av=0x606e88\, key=6319745\, val=0x6070b0\) at av\.c&#8203;:336
336            ary\[\+\+AvFILLp\(av\)\] = &PL\_sv\_undef;
\(gdb\) bt
\#0  0x00007ffff7d091a8 in Perl\_av\_store \(my\_perl=0x602010\, av=0x606e88\, key=6319745\, val=0x6070b0\) at av\.c&#8203;:336
\#1  0x00007ffff7d3f23a in Perl\_pp\_push \(my\_perl=0x602010\) at pp\.c&#8203;:4454
\#2  0x00007ffff7cead38 in Perl\_runops\_debug \(my\_perl=0x602010\) at dump\.c&#8203;:1931
\#3  0x00007ffff7d114f8 in S\_run\_body \(oldscope=\<value optimized out>\, my\_perl=\<value optimized out>\) at perl\.c&#8203;:2391
\#4  perl\_run \(oldscope=\<value optimized out>\, my\_perl=\<value optimized out>\) at perl\.c&#8203;:2309
\#5  0x0000000000400cfc in main \(argc=2\, argv=0x7fffffffe018\, env=0x7fffffffe030\) at perlmain\.c&#8203;:113

I am using Fedora 11\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl 5.10.0 in the Fedora build system.
It is being executed now by Perl 5.10.0 - Fri Apr 10 06:34:03 EDT 2009.

Site configuration information for perl 5.10.0:

Configured by Red Hat, Inc. at Fri Apr 10 06:34:03 EDT 2009.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.18-128.1.6.el5, archname=x86_64-linux-thread-multi
    uname='linux x86-3.fedora.phx.redhat.com 2.6.18-128.1.6.el5 #1 smp tue mar 24 12:05:57 edt 2009 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -DPERL_USE_SAFE_PUTENV -Dversion=5.10.0 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr/local -Dprivlib=/usr/lib/perl5/5.10.0 -Dsitelib=/usr/local/lib/perl5/site_perl/5.10.0 -Dvendorlib=/usr/lib/perl5/vendor_perl/5.10.0 -Darchlib=/usr/lib64/perl5/5.10.0/x86_64-linux-thread-multi -Dsitearch=/usr/local/lib64/perl5/site_perl/5.10.0/x86_64-linux-thread-multi -Dvendorarch=/usr/lib64/perl5/vendor_perl/5.10.0/x86_64-linux-thread-multi -Dinc_version_list=none -Darchname=x86_64-linux-thread-multi -Dlibpth=/usr/local/lib64 /lib64 /usr/lib64 -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly -Dpag!
 er=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin -Dotherlibdirs=/usr/lib/perl5/site_perl'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -DPERL_USE_SAFE_PUTENV',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -pipe -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='4.4.0 20090409 (Red Hat 4.4.0-0.32)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =''
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.9.90'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib64/perl5/5.10.0/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -DPERL_USE_SAFE_PUTENV'

Locally applied patches:
    


@INC for perl 5.10.0:
    /usr/local/ION/lib/perl
    /usr/local/ION/lib/perl
    /usr/local/lib64/perl5/site_perl/5.10.0/x86_64-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.10.0
    /usr/lib64/perl5/vendor_perl/5.10.0/x86_64-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.10.0
    /usr/lib/perl5/vendor_perl
    /usr/lib64/perl5/5.10.0/x86_64-linux-thread-multi
    /usr/lib/perl5/5.10.0
    /usr/lib/perl5/site_perl
    .


Environment for perl 5.10.0:
    HOME=/home/dstahlke
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/local/lib:/usr/local/jprofiler3/bin/linux-x86:/home/dstahlke/thesis/lib
    LOGDIR (unset)
    PATH=/usr/java/jdk1.6.0_04/bin:/usr/local/bin:/usr/java/jdk1.6.0_04/bin:/usr/local/bin:/usr/lib64/qt-3.3/bin:/usr/kerberos/bin:/usr/lib64/ccache:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/home/dstahlke/bin:/usr/local/Acrobat5/bin:/usr/local/ION/bin:/usr/java/jdk1.5.0_14/bin:/usr/local/jprofiler3/bin:/home/dstahlke/bin:/home/dstahlke/bin:/usr/local/Acrobat5/bin:/usr/local/ION/bin:/usr/java/jdk1.5.0_14/bin:/usr/local/jprofiler3/bin
    PERL5LIB=:/usr/local/ION/lib/perl:/usr/local/ION/lib/perl
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From nj88udd02@sneakemail.com

Hi all\,

dan@&#8203;gina\.alaska\.edu \(via RT\) wrote&#8203;:
> \# New Ticket Created by  dan@&#8203;gina\.alaska\.edu 
> \# Please include the string&#8203;:  \[perl \#66930\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=66930 >
\[\.\.\.\]

> The following program causes a segmentation fault&#8203;:
> 
> \#\!/usr/bin/perl 
> 
> use strict;
> use warnings;
> 
> my $x;
> \#my $x = \[\];  \# also crashes
> 
> push @&#8203;$x\, abc\(\);
> 
> sub abc \{
>     $x = 1;
>     \#$x = \[\];   \# also crashes
>     return 2;
>     \#return $x = 1;   \# also crashes
>     \# absense of return statement prevents crash
> \}

Confirmed with blead at 3244086bce52e956e04c61b12eb7b67fc1041d8b\.

Cheers\,
Steffen

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From perl@profvince.com

This happens because the array referenced by $x is pushed on the stack\,
and then is freed when abc\(\) resets $x\. Hence push tries to access a
freed SV\. It can be fixed by deciding at compile time to pass the
reference itself and letting push\(\) dereference it\, after the
interpreter is done with the argument list\.

Note that replacing push by a prototyped sub doesn't segfault :

&nbsp;   sub def \(\\@&#8203;$\) \{ print @&#8203;\{$\_\[0\]\} \}
&nbsp;   my $x;
&nbsp;   def @&#8203;$x\, do \{ $x = \[ 1 \]; 42 \};

This will print "1"\, because the first argument is actually a new
reference to @&#8203;$x\, so it survives even if $x is changed\. On a side note\,
I'm unsure whether this is a very nice behaviour : I'd expect the first
pushed reference to be the one I have access to\. Think :

&nbsp;   our $global;
&nbsp;   def @&#8203;$global\, opaque\_sub\_that\_modifies\_global\_but\_you\_dont\_know\_it\(\);

In any case\, it's not consistent with push\(\) where you pass the array
its original state \(even if currently it crashes\)\.


Vincent\.

**Comment by p5pRT** [comment]
### From @davidnicol

I shrunk the demonstration a little for addition to t/op/list\.t\.
Maybe push could be made to really have a \\@&#8203;@&#8203; prototype instead of
whatever it does now?  Where that subroutine result winds up getting
pushed to is not the question here\, but I think left to right
processing implies it gets pushed to whatever array $x referred to
before calling the subroutine\, so making that new reference is needed
because the reference in the argument can't be trusted to still be an
array reference when we get to the pushing part\.  When push comes to
shove as it were\.  Take it away\, Vincent Pit\!

<details><summary>Inline Patch</summary>

```diff
diff --git a/t/op/list.t b/t/op/list.t
index a8fdc18..b678224 100644
--- a/t/op/list.t
+++ b/t/op/list.t
@@ -6,7 +6,7 @@ BEGIN {
 }

 require "test.pl";
-plan( tests => 57 );
+plan( tests => 59 );

 @foo = (1, 2, 3, 4);
 cmp_ok($foo[0], '==', 1, 'first elem');
@@ -161,3 +161,21 @@ cmp_ok(join('',(1,2),3,(4,5)),'eq','12345','list

```
</details>
(..).(..)');
&nbsp;     test\_zero\_args\("do\-returned list slice"\, do \{ \(10\,11\)\[2\,3\]; \}\);
 \}

\+\{
\+    \# perl \#66930
\+    fresh\_perl\_is \(\<\<'WORKS'\,"cheese"\,\{\}\,"66930 push segfault null case"\);
\+push @&#8203;$x\, \(sub \{
\+    $x = 'cheese';
\+\}\)\->\(\);
\+print "$x\\n";
\+WORKS
\+    fresh\_perl\_is \(\<\<'66930'\,"cheese"\,\{\}\,"66930 push segfault"\);
\+push @&#8203;$x\, \(sub \{
\+    $x = 'cheese';
\+    2;
\+\}\)\->\(\);
\+print "$x\\n";
\+66930
\+
\+\}
\+

**Comment by p5pRT** [comment]
### From @davidnicol

>> Where that subroutine result winds up getting
>> pushed to is not the question here\, but I think left to right
>> processing implies it gets pushed to whatever array $x referred to
>> before calling the subroutine

On the other hand\, that subroutine arguments are all aliases and that
all arguments are evaluated before the subroutine is invoked would put
the subroutine result in the later array\.  Where does the
non\-seg\-faulting version put it?  In the earlier array\.

With the \\@&#8203;@&#8203; prototype it goes in the earlier array too\.  With the
implied \(@&#8203;\) prototype\, the mypush2 routine below puts "shove" in the
after array\.

 cat 66930\_where\_does\_it\_go\.pl
@&#8203;before=@&#8203;after=\(\);
$x="before";
push @&#8203;$x\, \(sub \{
&nbsp;       $x = 'after';
\}\)\->\(\);
print "PURE PUSH&#8203;:\\nx&#8203;: $x\\nbefore&#8203;:@&#8203;before\\nafter&#8203;: @&#8203;after\\n";

sub mypush\(\\@&#8203;@&#8203;\) \{ my $ref = shift; push @&#8203;$ref\, @&#8203;\_ \}
@&#8203;before=@&#8203;after=\(\);
$x="before";
mypush @&#8203;$x\, \(sub \{
&nbsp;       $x = 'after';
\}\)\->\(\);
print "MY PUSH&#8203;:\\nx&#8203;: $x\\nbefore&#8203;:@&#8203;before\\nafter&#8203;: @&#8203;after\\n";

@&#8203;before=@&#8203;after=\(\);
$x="before";
mypush @&#8203;$x\, \(sub \{
&nbsp;       $x = 'after';
&nbsp;       "shove"
\}\)\->\(\);
print "WOULD SEGFAULT&#8203;:\\nx&#8203;: $x\\nbefore&#8203;:@&#8203;before\\nafter&#8203;: @&#8203;after\\n";

@&#8203;before=@&#8203;after=\(\);
$x="before";
$sub\_result = \(sub \{ $x = 'after'; "shove" \}\)\->\(\);
push @&#8203;$x\, $sub\_result;
print "PERL\-SIDE ORDERING&#8203;:\\nx&#8203;: $x\\nbefore&#8203;:@&#8203;before\\nafter&#8203;: @&#8203;after\\n";

@&#8203;before=@&#8203;after=\(\);
$x="before";
$arrayref = $x;
$sub\_result = \(sub \{ $x = 'after'; "shove" \}\)\->\(\);
push @&#8203;$arrayref\, $sub\_result;
print "PERL\-SIDE ORDERING 2&#8203;:\\nx&#8203;: $x\\nbefore&#8203;:@&#8203;before\\nafter&#8203;: @&#8203;after\\n";

sub mypush2 \{ my $ref = shift; push @&#8203;$ref\, @&#8203;\_ \}
@&#8203;before=@&#8203;after=\(\);
$x="before";
mypush2 $x\, \(sub \{
&nbsp;       $x = 'after';
&nbsp;       "shove"
\}\)\->\(\);
print "MY PUSH NO PROTOTYPE&#8203;:\\nx&#8203;: $x\\nbefore&#8203;:@&#8203;before\\nafter&#8203;: @&#8203;after\\n";
\_\_END\_\_

**Comment by p5pRT** [comment]
### From @davidnicol

Stunningly\, unshift doesn't have the segfault problem\. Like push that
works\, or a \\@&#8203;\.\.\. prototype\, the array that gets modified is as seen
before rather than after the modification in the subroutine\.

I wonder if a note about this edge case of parameter ordering being
another effect of prototyping belongs in perlsub\.pod\.

pp\_push and pp\_unshift in pp\.c are very similar\.

Hmm\.  5\.10\.0 \(standard cygwin\) segfaults on push but not unshift\.

5\.8\.8 \(debian package\) segfaults on both\.

Current blead on linux allows push\, complains and aborts on unshift\,
and I have no idea why\.  The problem is in the returning\, not in the
operations\. The time of destruction has been deferred for push and the
same deferment needs to be applied to unshift too?

<details><summary>Inline Patch</summary>

```diff
diff --git a/pp.c b/pp.c
index 9cedc3f..9318755 100644
--- a/pp.c
+++ b/pp.c
@@ -4549,6 +4549,7 @@ PP(pp_push)
        }
     }
     else {
+           Perl_warn(aTHX_ "inside push refcount %i", SvREFCNT((SV*) ary));
        PL_delaymagic = DM_DELAY;
        for (++MARK; MARK <= SP; MARK++) {
            SV * const sv = newSV(0);
@@ -4596,17 +4597,23 @@ PP(pp_unshift)
        SPAGAIN;
     }
     else {
+           Perl_warn(aTHX_ "inside unshift refcount %i", SvREFCNT((SV*) ary));
        register I32 i = 0;
        av_unshift(ary, SP - MARK);
+           Perl_warn(aTHX_ "unshifted %i", SP - MARK);
        while (MARK < SP) {
+           Perl_warn(aTHX_ "unshifting %i", SP - MARK);
            SV * const sv = newSVsv(*++MARK);
+           Perl_warn(aTHX_ "unshifting %i", SP - MARK);
            (void)av_store(ary, i++, sv);
+           Perl_warn(aTHX_ "unshifting %i", SP - MARK);
        }
     }
     SP = ORIGMARK;
     if (GIMME_V != G_VOID) {
        PUSHi( AvFILL(ary) + 1 );
     }
+           Perl_warn(aTHX_ "returning");
     RETURN;
 }

diff --git a/utils/Makefile b/utils/Makefile



```
</details>
perlhacker@catnip:~/perl$ ./perl -v

This is perl\, v5\.11\.0 \(GitLive\-blead\-g7f6de3f\*\) built for i686\-linux
\.\.\.

perlhacker@&#8203;catnip&#8203;:\~/perl$ \./perl \-le '$x=\[\]; push @&#8203;$x\, \(sub \{$x="";"pbth"\}\)\->\(\)'
inside push refcount 1 at \-e line 1\.
perlhacker@&#8203;catnip&#8203;:\~/perl$ \./perl \-le '$x=\[\]; unshift @&#8203;$x\, \(sub
\{$x="";"pbth"\}\)\->\(\)'
inside unshift refcount 1 at \-e line 1\.
unshifted 1 at \-e line 1\.
unshifting 1 at \-e line 1\.
unshifting 0 at \-e line 1\.
unshifting 0 at \-e line 1\.
returning at \-e line 1\.
\*\*\* glibc detected \*\*\* free\(\)&#8203;: invalid pointer&#8203;: 0x0816939c \*\*\*
Aborted

**Comment by p5pRT** [comment]
### From @iabyn

Without looking at this thread in detail \(I just haven't got time at the
moment\)\, my general observation is that most of these bugs are due to
the fact that the perl stack isn't refcounted\, which is a fairly
fundamental problem with no quick fixes\. So the usual advice is just
"don't modify an array while you're iterating over it / using it as a
function arg / etc etc; and maybe well fix this one year"\.


On Sun\, Jun 28\, 2009 at 07&#8203;:59&#8203;:48PM \-0500\, David Nicol wrote&#8203;:
> Stunningly\, unshift doesn't have the segfault problem\. Like push that
> works\, or a \\@&#8203;\.\.\. prototype\, the array that gets modified is as seen
> before rather than after the modification in the subroutine\.
> 
> I wonder if a note about this edge case of parameter ordering being
> another effect of prototyping belongs in perlsub\.pod\.
> 
> pp\_push and pp\_unshift in pp\.c are very similar\.
> 
> Hmm\.  5\.10\.0 \(standard cygwin\) segfaults on push but not unshift\.
> 
> 5\.8\.8 \(debian package\) segfaults on both\.
> 
> Current blead on linux allows push\, complains and aborts on unshift\,
> and I have no idea why\.  The problem is in the returning\, not in the
> operations\. The time of destruction has been deferred for push and the
> same deferment needs to be applied to unshift too?
> 
> diff \-\-git a/pp\.c b/pp\.c
> index 9cedc3f\.\.9318755 100644
> \-\-\- a/pp\.c
> \+\+\+ b/pp\.c
> @&#8203;@&#8203; \-4549\,6 \+4549\,7 @&#8203;@&#8203; PP\(pp\_push\)
>         \}
>      \}
>      else \{
> \+           Perl\_warn\(aTHX\_ "inside push refcount %i"\, SvREFCNT\(\(SV\*\) ary\)\);
>         PL\_delaymagic = DM\_DELAY;
>         for \(\+\+MARK; MARK \<= SP; MARK\+\+\) \{
>             SV \* const sv = newSV\(0\);
> @&#8203;@&#8203; \-4596\,17 \+4597\,23 @&#8203;@&#8203; PP\(pp\_unshift\)
>         SPAGAIN;
>      \}
>      else \{
> \+           Perl\_warn\(aTHX\_ "inside unshift refcount %i"\, SvREFCNT\(\(SV\*\) ary\)\);
>         register I32 i = 0;
>         av\_unshift\(ary\, SP \- MARK\);
> \+           Perl\_warn\(aTHX\_ "unshifted %i"\, SP \- MARK\);
>         while \(MARK \< SP\) \{
> \+           Perl\_warn\(aTHX\_ "unshifting %i"\, SP \- MARK\);
>             SV \* const sv = newSVsv\(\*\+\+MARK\);
> \+           Perl\_warn\(aTHX\_ "unshifting %i"\, SP \- MARK\);
>             \(void\)av\_store\(ary\, i\+\+\, sv\);
> \+           Perl\_warn\(aTHX\_ "unshifting %i"\, SP \- MARK\);
>         \}
>      \}
>      SP = ORIGMARK;
>      if \(GIMME\_V \!= G\_VOID\) \{
>         PUSHi\( AvFILL\(ary\) \+ 1 \);
>      \}
> \+           Perl\_warn\(aTHX\_ "returning"\);
>      RETURN;
>  \}
> 
> diff \-\-git a/utils/Makefile b/utils/Makefile
> 
> 
> perlhacker@&#8203;catnip&#8203;:\~/perl$ \./perl \-v
> 
> This is perl\, v5\.11\.0 \(GitLive\-blead\-g7f6de3f\*\) built for i686\-linux
> \.\.\.
> 
> perlhacker@&#8203;catnip&#8203;:\~/perl$ \./perl \-le '$x=\[\]; push @&#8203;$x\, \(sub \{$x="";"pbth"\}\)\->\(\)'
> inside push refcount 1 at \-e line 1\.
> perlhacker@&#8203;catnip&#8203;:\~/perl$ \./perl \-le '$x=\[\]; unshift @&#8203;$x\, \(sub
> \{$x="";"pbth"\}\)\->\(\)'
> inside unshift refcount 1 at \-e line 1\.
> unshifted 1 at \-e line 1\.
> unshifting 1 at \-e line 1\.
> unshifting 0 at \-e line 1\.
> unshifting 0 at \-e line 1\.
> returning at \-e line 1\.
> \*\*\* glibc detected \*\*\* free\(\)&#8203;: invalid pointer&#8203;: 0x0816939c \*\*\*
> Aborted

\-\- 
The warp engines start playing up a bit\, but seem to sort themselves out
after a while without any intervention from boy genius Wesley Crusher\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#17

**Comment by p5pRT** [comment]
### From @davidnicol

On Tue\, Jun 30\, 2009 at 7&#8203;:54 AM\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:

> Without looking at this thread in detail \(I just haven't got time at the
> moment\)\, my general observation is that most of these bugs are due to
> the fact that the perl stack isn't refcounted\, which is a fairly
> fundamental problem with no quick fixes\. So the usual advice is just
> "don't modify an array while you're iterating over it / using it as a
> function arg / etc etc; and maybe well fix this one year"\.


While it is true that this bug is clearly of the "so don't do that" variety\,
a segfault in any circumstance seems like spilled milk worth wiping up\.

While a refcounted stack might prevent freeing the no\-longer\-referenced
array too early\, whatever method is used to call an explicitly prototyped
userland routine manages to escape without segfaulting\. That push and
unshift appear to have switched status as to if they segfault or not is
intruiging\.

Don't we want to normalize the builtins to use the prototyped calling
conventions?  I thought at one point we did; that initiative like so many
others may have fizzled or been rejected due to a slight benchmarked
performance hit\.

**Comment by p5pRT** [comment]
### From @jimc

> Don't we want to normalize the builtins to use the prototyped calling
> conventions?  I thought at one point we did; that initiative like so many
> others may have fizzled or been rejected due to a slight benchmarked
> performance hit\.
>


Id suspect that its more like&#8203;:

These are problems that are understood\, if not by everyone\,
then by enough who can say with some authority&#8203;: "dont do that"\.

Im unconvinced that fixing 70% of the bugs I dont understand
will help me at all when I find myself blocked by the other 30%\.

Wisdom \(self\-preservation?\) says stay the hell out of the minefield\,
It only takes one to go boom\, stumpy\.

Then theres that "we" in there\.
Only the queen gets to use that word\.
\(and Im as guilty of that as anyone\)

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Jul 02\, 2009 at 07&#8203;:22&#8203;:45PM \-0500\, David Nicol wrote&#8203;:
> While a refcounted stack might prevent freeing the no\-longer\-referenced
> array too early\, whatever method is used to call an explicitly prototyped
> userland routine manages to escape without segfaulting\. That push and
> unshift appear to have switched status as to if they segfault or not is
> intruiging\.

You appear to be conflating "it doesn't segfault" with "there's nothing
wrong"\. An underlying recount bug may cause some operations to segfault
but not others\, and what does and doesn't segfault may vary randomly
between versions\.

Anyway\, as I said before\, I haven't got time to discuss this now\.

\-\- 
Little fly\, thy summer's play my thoughtless hand
has terminated with extreme prejudice\.
&nbsp;        \(with apologies to William Blake\)
